src/arcaflow_plugin_sdk/schema.py

```diff
@@ -5568,51 +5686,91 @@ class SchemaType(Schema):
+        """
+        This function unserializes the input from a raw data to data structures, such as dataclasses. This function is
+        automatically called by ``__call__`` before running the step with the unserialized input.
+
+        :param step_id: The step ID to use to look up the schema for unserialization.
+        :param serialized_data: The raw data to unserialize.
+        :return: The unserialized data in the structure the step expects it.
+        """
+        return self._unserialize_signal_handler_input(self.get_signal(step_id, signal_id), serialized_data)
 
     @staticmethod
-    def _unserialize_input(step: StepType, data: Any) -> Any:
+    def _unserialize_signal_handler_input(signal: SignalHandlerType, data: Any) -> Any:
         try:
-            return step.input.unserialize(data)
+            return signal.data_schema.unserialize(data)
         except ConstraintException as e:
             raise InvalidInputException(e) from e
 
     def call_step(self, step_id: str, input_param: Any) -> typing.Tuple[str, Any]:
         """
         This function calls a specific step with the input parameter that has already been unserialized. It expects the
-        data to be already valid, use unserialize_input to produce a valid input. This function is automatically called
-        by ``__call__`` after unserializing the input.
+        data to be already valid, use unserialize_step_input to produce a valid input. This function is automatically
+        called by ``__call__`` after unserializing the input.
 
         :param step_id: The ID of the input step to run.
         :param input_param: The unserialized data structure the step expects.
         :return: The ID of the output, and the data structure returned from the step.
         """
-        if step_id not in self.steps:
-            raise NoSuchStepException(step_id)
-        step = self.steps[step_id]
-        return self._call_step(step, input_param)
+        return self._call_step(self.get_step(step_id), input_param)
+
+    def call_step_signal(self, step_id: str, signal_id: str, unserialized_input_param: Any):
+        """
+        This function calls a specific step's signal with the input parameter that has already been unserialized. It expects the
+        data to be already valid, use unserialize_signal_input to produce a valid input.
+
+        :param step_id: The ID of the input step to run.
+        :param unserialized_input_param: The unserialized data structure the step expects.
+        :return: The ID of the output, and the data structure returned from the step.
+        """
+        step = self.get_step(step_id)
+        signal = self.get_signal(step_id, signal_id)
+        if not step.object_data_ready:
+            with step.object_cv:
+                # wait to be notified of it being ready. Test this by adding a sleep before the step call.
+                step.object_cv.wait()
```

Comment:
This code looks questionable:  it probably should hold the `object_cv` lock when checking the `object_data_ready` value, and it should be prepared for spurious wake-ups.  So, something more like this would be appropriate:
```python
        with step.object_cv:
            while not step.object_data_ready:
                step.object_cv.wait()
```
(I would omit the code comment:  it's obvious that we are waiting, it's clear what we are waiting for, and comments on how to test this belong in the test module not here....)